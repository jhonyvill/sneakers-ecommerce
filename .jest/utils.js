import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../src/Themes";

export function renderWithTheme(children) {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
}

export const productDefault = {
  id: "id-1",
  brand: "brand mock",
  title: "title mock",
  price: 250,
  discount: 0.5,
  description: "description mock",
  pictures: [
    {
      src: "image-1.jpg",
      thumb: "thumbnail-1.jpg",
      alt: "product 1",
    },
    {
      src: "image-2.jpg",
      thumb: "thumbnail-2.jpg",
      alt: "product 2",
    },
    {
      src: "image-3.jpg",
      thumb: "thumbnail-3.jpg",
      alt: "product 3",
    },
    {
      src: "image-4.jpg",
      thumb: "thumbnail-4.jpg",
      alt: "product 4",
    },
    {
      src: "image-5.jpg",
      thumb: "thumbnail-5.jpg",
      alt: "product 5",
    },
  ],
};

export const itemDefault = {
  product: { ...productDefault },
  quantity: 1,
};
