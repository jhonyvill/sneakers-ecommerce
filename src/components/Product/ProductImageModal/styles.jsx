import styled, { css } from "styled-components";

export const Container = styled.section`
  position: fixed;
  width: 100%;
  height: 100vh;
  background-color: ${({ theme }) => theme.colors.blackOpacity};
  left: 0;
  top: 0;
  z-index: 200;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const Modal = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
`;

export const ImageContainer = styled.div`
  display: flex;
  align-items: center;
  position: relative;
  width: 30rem;
  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    width: 100%;
  }
`;

export const CloseButton = styled.button`
  position: absolute;
  top: -2rem;
  right: -0.5rem;
  color: ${({ theme }) => theme.colors.lightGrayishBlue};
  background-color: transparent;
  border: 1px solid transparent;

  & svg > * {
    fill: ${({ theme }) => theme.colors.lightGrayishBlue};
  }
  &:hover svg > * {
    fill: ${({ theme }) => theme.colors.primaryOrange};
  }
`;
