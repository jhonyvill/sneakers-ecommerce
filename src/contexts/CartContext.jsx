import React from "react";
import { useReducer } from "react";

export const CartContext = React.createContext();

export const CartProvider = ({ children }) => {
  const initialState = { items: [] };

  function cartReducer(state, action) {
    switch (action.type) {
      case "ADD_ITEM":
        return { ...state, items: [...state.items, action.payload] };
      case "UPDATE_ITEM":
        const { product } = action.payload;
        const itemIndex = state.items.findIndex(
          (item) => item.product.id === product.id
        );

        const updatedItems = [...state.items];
        updatedItems.slice(itemIndex, product);
        return { ...state, items: updatedItems };
      default:
        state;
    }
  }

  const [state, dispatch] = useReducer(cartReducer, initialState);

  return (
    <>
      <CartContext.Provider
        value={{ cartState: state, cartDispatch: dispatch }}
      >
        {children}
      </CartContext.Provider>
    </>
  );
};
