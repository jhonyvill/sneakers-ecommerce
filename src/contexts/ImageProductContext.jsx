import React from "react";
import { useState } from "react";

export const ImageProductContext = React.createContext();

export const ImageProductProvider = ({ children }) => {
  const [selectedImage, setSelectedImage] = useState({
    image: null,
    index: 0,
  });

  const handleSelectedModalImage = (image, index) => {
    setSelectedImage({ image, index });
  };

  return (
    <>
      <ImageProductContext.Provider
        value={{
          imageSelected: selectedImage.image,
          indexImageSelected: selectedImage.index,
          handleSelectedModalImage,
        }}
      >
        {children}
      </ImageProductContext.Provider>
    </>
  );
};
