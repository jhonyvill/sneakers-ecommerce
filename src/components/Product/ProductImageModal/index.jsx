import React from "react";
import * as S from "./styles";
import CloseSvg from "../../../assets/icon-close.svg?react";
import PropTypes from "prop-types";
import ProductImages from "../ProductImages";

const ProductImageModal = ({ images, handleModalActive }) => {
  return (
    <S.Container aria-label="modal imagens do produto">
      <S.Modal>
        <S.CloseButton data-testid='close-button' onClick={() => handleModalActive(false)}>
          <CloseSvg />
        </S.CloseButton>

        <S.ImageContainer>
          <ProductImages
            images={images}
            modalActive={true}
            handleModalActive={handleModalActive}
          />
        </S.ImageContainer>
      </S.Modal>
    </S.Container>
  );
};

ProductImageModal.protoTypes = {
  images: PropTypes.array.isRequired,
  handleModalActive: PropTypes.func.isRequired,
};

export default ProductImageModal;
