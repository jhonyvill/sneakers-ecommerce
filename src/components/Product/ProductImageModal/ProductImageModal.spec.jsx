import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme, productDefault } from "../../../../.jest/utils";
import { ImageProductProvider } from "../../../contexts/ImageProductContext";
import ProductImageModal from "./";

function renderProductImageModal(images, handleModalActive = jest.fn()){
  return renderWithTheme(
    <ImageProductProvider>
      <ProductImageModal images={images} handleModalActive={handleModalActive} />
    </ImageProductProvider>
  );
}

jest.mock("../../../hooks/useMedia", () => ({
  __esModule: true,
  useMedia: () => {
    return false;
  },
}));

const images = productDefault.pictures;

describe("<ProductImageModal />", () => {
  it("should render close button and images", async () => {
    renderProductImageModal(images);

    const closeModalButton = screen.getByTestId("close-button");
    const featuredImage = screen.getByTestId("featured-image");
    const regionProductImages = screen.getByRole('region', { name: "imagens do produto" });
    const gridImages = screen.getAllByTestId(/grid-image/i);
    const previousButton = screen.getByRole("button", { name: /imagem anterior/i });
    const nextButton = screen.getByRole("button", { name: /prÃ³xima imagem/i });

    expect(regionProductImages).toBeInTheDocument();
    expect(closeModalButton).toBeInTheDocument();
    expect(previousButton).toBeInTheDocument();
    expect(featuredImage).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();
    expect(gridImages).toHaveLength(4)
  });

  it("should call function on click modal close button", async () => {
    const mockedFunction = jest.fn();
    renderProductImageModal(images, mockedFunction);

    const closeModalButton = screen.getByTestId("close-button");
    await userEvent.click(closeModalButton);

    expect.assertions(1);
    expect(mockedFunction).toHaveBeenCalledTimes(1);
  });
});

describe("<ProductImageModal /> snapshots", () => {
  it("should match snapshot", () => {
    const { container } = renderProductImageModal(images);
    expect(container.firstChild).toMatchSnapshot();
  });
});


