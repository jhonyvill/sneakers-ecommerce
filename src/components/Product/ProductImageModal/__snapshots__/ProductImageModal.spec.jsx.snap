// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductImageModal /> snapshots should match snapshot 1`] = `
<section
  aria-label="modal imagens do produto"
  class="styles__Container-sc-tc5yrj-0 iYchpi"
>
  <div
    class="styles__Modal-sc-tc5yrj-1 fuToiU"
  >
    <button
      class="styles__CloseButton-sc-tc5yrj-3 fscaYl"
      data-testid="close-button"
    >
      <icon-mock />
    </button>
    <div
      class="styles__ImageContainer-sc-tc5yrj-2 hlTyuA"
    >
      <section
        aria-label="imagens do produto"
        class="styles__Container-sc-1rfz8de-0 cjexoL"
      >
        <div
          class="styles__FeaturedImageContainer-sc-1rfz8de-1 hkalnS"
          data-testid="container-featured-image"
        >
          <img
            alt="product 1"
            class="styles__FeaturedImage-sc-1rfz8de-2 hLfEin"
            data-testid="featured-image"
            height="1000px"
            src="image-1.jpg"
            width="1000px"
          />
          <div
            class="styles__ButtonWrapper-sc-1rfz8de-5 hFyXwb"
          >
            <button
              aria-label="imagem anterior"
              class="styles__ModalButton-sc-1rfz8de-6 styles__PreviousButton-sc-1rfz8de-7 bobShG kxWovx"
            >
              <icon-mock />
            </button>
            <button
              aria-label="prÃ³xima imagem"
              class="styles__ModalButton-sc-1rfz8de-6 styles__NextButton-sc-1rfz8de-8 bobShG gUtcMk"
            >
              <icon-mock />
            </button>
          </div>
        </div>
        <div
          class="styles__ImagesGrid-sc-1rfz8de-3 gtpYQD"
        >
          <div
            class="styles__Image-sc-1rfz8de-4 gGfERl active"
            data-testid="grid-image-0"
          >
            <img
              alt="product 1"
              height="176px"
              src="thumbnail-1.jpg"
              width="176px"
            />
          </div>
          <div
            class="styles__Image-sc-1rfz8de-4 gGfERl"
            data-testid="grid-image-1"
          >
            <img
              alt="product 2"
              height="176px"
              src="thumbnail-2.jpg"
              width="176px"
            />
          </div>
          <div
            class="styles__Image-sc-1rfz8de-4 gGfERl"
            data-testid="grid-image-2"
          >
            <img
              alt="product 3"
              height="176px"
              src="thumbnail-3.jpg"
              width="176px"
            />
          </div>
          <div
            class="styles__Image-sc-1rfz8de-4 gGfERl"
            data-testid="grid-image-3"
          >
            <img
              alt="product 4"
              height="176px"
              src="thumbnail-4.jpg"
              width="176px"
            />
          </div>
        </div>
      </section>
    </div>
  </div>
</section>
`;
